logger/buffered_logger.go

```diff
@@ -98,166 +98,178 @@ func (bl *bufferedLogger) Start(
 	cleanupTime *time.Duration,
 	ready func() error,
 ) error {
-	stdout, stderr := bl.l.GetPipes()
-	if stdout == nil || stderr == nil {
-		return errors.New("no stdout/stderr pipe opened")
+	pipeNameToPipe, err := bl.l.GetPipes()
+	if err != nil {
+		return err
 	}
 
-	var wg sync.WaitGroup
-	wg.Add(1)
-	debug.SendEventsToJournal(DaemonName,
-		"Starting reading from stdout pipe",
-		journal.PriInfo, 0)
-	go bl.saveLogsToBuffer(stdout, &wg, sourceSTDOUT, uid, gid)
-
-	wg.Add(1)
-	debug.SendEventsToJournal(DaemonName,
-		"Starting reading from stderr pipe",
-		journal.PriInfo, 0)
-	go bl.saveLogsToBuffer(stderr, &wg, sourceSTDERR, uid, gid)
+	errGroup, ctx := errgroup.WithContext(ctx)
+	// Start the goroutine of underling log driver to consume logs from ring buffer and
+	// send logs to destination when there's any.
+	errGroup.Go(func() error {
+		debug.SendEventsToJournal(DaemonName, "Starting consuming logs from ring buffer", journal.PriInfo, 0)
+		return bl.sendLogMessagesToDestination(uid, gid, cleanupTime)
+	})
+
+	// Start reading logs from container pipes.
+	for pn, p := range pipeNameToPipe {
+		// Copy pn and p to new variables source and pipe, accordingly.
+		source := pn
+		pipe := p
+
+		errGroup.Go(func() error {
+			logErr := bl.saveLogMessagesToRingBuffer(ctx, pipe, source, uid, gid)
+			if logErr != nil {
+				err := errors.Wrapf(logErr, "failed to send logs from pipe %s", source)
+				debug.SendEventsToJournal(DaemonName, err.Error(), journal.PriErr, 1)
+				return err
+			}
+			return nil
+		})
+	}
 
 	// Signal that the container is ready to be started
 	if err := ready(); err != nil {
-		return errors.Wrap(err, "failed to notify container ready status to containerd")
+		return errors.Wrap(err, "failed to check container ready status")
 	}
 
-	// Start the underling log driver to send logs to destination
-	wg.Add(1)
-	go bl.sendLogs(&wg, uid, gid, cleanupTime)
-
-	wg.Wait()
-	debug.SendEventsToJournal(DaemonName,
-		"All logs saved to buffer has been sent to destination",
-		journal.PriInfo, 1)
-
-	return nil
+	// Wait() will return the first error it receives.
+	return errGroup.Wait()
 }
 
-// saveLogsToBuffer saves container logs to intermediate buffer.
-func (bl *bufferedLogger) saveLogsToBuffer(f io.Reader, wg *sync.WaitGroup, source string, uid int, gid int) {
-	defer wg.Done()
-
+// saveLogMessagesToRingBuffer saves container log messages to ring buffer.
+func (bl *bufferedLogger) saveLogMessagesToRingBuffer(
```

Comment:
Can we differentiate this method and `saveLogMsgToRingBuffer` somehow? The names of these methods are very similar and can get confusing.

Reply:
What about `saveLogMessagesToRingBuffer` and `saveSingleLogMessageToRingBuffer`?

Reply:
That works for me, feel free to push once the changes are made, thanks!

Reply:
Thanks!